### ISSUE TITLE
Automatically provision Rancher agents with Terraform without cycle

### ISSUE THREAD

**Issue by mattdodge** [rancher/rancher#9764]
**Rancher versions:**
rancher/server: 1.5.3
rancher/agent: 1.2.1

This is more of a question of whether this is possible or not. We are using Terraform and Ansible to manage our cloud infrastructure. I would like new EC2 instances to be automatically registered with the Rancher cluster by running the agent. However, these are EC2 hosts behind a NAT so we need to explicitly set the agent IP for the Rancher agent. But this creates a cyclical dependency between the `rancher_registration_token` and `aws_instance` resources.

```tf
resource "rancher_registration_token" "api_token" {
  name = "${var.env}-api_token"
  environment_id = "${rancher_environment.api.id}"
  host_labels = {
    "role" = "API HOST"
  }
  agent_ip = "${aws_instance.api.private_ip)}"
}

resource "aws_instance" "api" {
  ami = "${var.ami}"
  provisioner "local-exec" {
    command = <<EOF
      ansible-playbook \
      -i '${self.private_ip},' \
      -e "rancher_command=\"${rancher_registration_token.api_token.command}\"" \
      ansible/playbooks/host.playbook.yml;
    EOF
  }
}
```

Notice that the registration token resource needs the IP of the instance but the instance's provisioner needs the command from the registration token. Is there any way to do this without having this cycle occur?

**Comment by superseb** [comment]
Please see this example on how to handle this kind of situation: https://github.com/rancher/terraform-modules/blob/master/example_ha/aws/kubernetes-env/files/userdata.template

Shouldn't be too hard to port this from user-data to your playbook and get the `export CATTLE_AGENT_IP` exported in side `/var/lib/rancher/etc/agent.conf`

Let me know if you run into trouble.

**Comment by mattdodge** [comment]
Hmm I see, that's an interesting approach. Write to `agent.conf` before running the rancher agent container.

We ended up not using the `command` output of the `rancher_registration_token` module and instead just used the `token` and `registration_url` and built the docker run command inside the playbook. The downside of that though is that you don't get to leverage the `host_labels` in the module definition. 

We were able to make this work but I like the `agent.conf` approach a bit better. I'll give that a go and report back if anything goes whacky. Otherwise, I think this is a viable solution, thanks again.
